### PR TITLE
Main page #16: 

### PR DESCRIPTION
**Pull request by myroslav-kolomiiets**
made base structure of main page, with list of popular movies outputted by default and simple search

### CODE REVIEW

File: src/components/MainPage.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import React, { Component } from 'react';
+import '../App.css';
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Pls, use absolute paths. They easier to understand and to reuse.

File: src/components/MainPage.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    this.state = {
+      error: null,
+      isLoaded: false,
+      items: [],
+      inputValue: '',
+    };
+  }
+
+  componentDidMount() {
+    fetch(
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Pls, create file `api.js` and move such functionality there

File: src/components/MainPage.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
+import React, { Component } from 'react';
```
Comment (zonzujiro) [inline src/components/MainPage.js:2]: You forgot `flow` comment?

File: src/components/MainPage.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    };
+  }
+
+  componentDidMount() {
+    fetch(
+      'https://api.themoviedb.org/3/discover/movie?api_key=5265606bb69e99437c85eb04dc6f29b5&sort_by=popularity.desc'
+    )
+      .then(res => res.json())
+      .then(
+        result => {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Such stuff must be handled thru redux modules

File: src/components/MainPage.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+          this.setState({
+            isLoaded: true,
+            items: result.results,
+          });
+          console.log(this.state.items);
```
Comment (zonzujiro) [inline src/components/MainPage.js]: 👀

File: src/components/MainPage.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  render() {
+    const { error, isLoaded, items } = this.state;
+    const poster = 'https://image.tmdb.org/t/p/w500';
+    if (error) {
+      return <div>Error: {error.message}</div>;
+    } else if (!isLoaded) {
+      return <div>Loading...</div>;
+    } else {
+      return (
+        <div>
```
Comment (zonzujiro) [inline src/components/MainPage.js]: `<></>`

**Review by zonzujiro — CHANGES_REQUESTED** [review]

File: package.json [hunk 7]
```diff
@@ -7,6 +7,7 @@
     "@material-ui/core": "^1.2.3",
     "@material-ui/icons": "^1.1.0",
     "lodash": "^4.17.10",
+    "prop-types": "^15.6.2",
```
Comment (zonzujiro) [inline package.json]: You don't need `prop-types` because you have `Flow`. Pls, remove and use types instead

File: src/App.test.js [hunk 8]
```diff
@@ -1,9 +1,13 @@
 import React from 'react';
 import ReactDOM from 'react-dom';
 import App from './App';
+import configureStore from './store/configureStore';
+import {Provider} from 'react-redux';
```
Comment (zonzujiro) [inline src/App.test.js]: Looks like `prettier` not worked

File: src/actions/actionTypes.js [hunk 9]
```diff
@@ -0,0 +1,4 @@
+/* @flow */
+export const MOVIE_LOADING = 'MOVIE_LOADING';
```
Comment (zonzujiro) [inline src/actions/actionTypes.js]: According to ducks-modules action types must be in the module

File: src/actions/actions.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
+/* @flow */
```
Comment (zonzujiro) [inline src/actions/actions.js]: Move this to the module.

Also, I think that you should reuse class made by @OleksiyRudenko for API calls

File: src/actions/actions.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  return {type: types.MOVIE_LOADED, data: json.results};
+}
+
+export function receiveSearchMovies(json) {
+  return {type: types.MOVIE_SEARCH, data: json.results};
+}
+
+export function fetchMovies() {
+  return dispatch => {
+    return fetch(`${BASE_URL}${TOP_FILMS}?${API_KEY}&${SORT_BY_POPULARITY}`)
```
Comment (zonzujiro) [inline src/actions/actions.js]: Can you use `async/await` here? The code will become more understandable

File: src/components/MainPage.js [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+import CardActions from '@material-ui/core/CardActions';
+import CardContent from '@material-ui/core/CardContent';
+import CardMedia from '@material-ui/core/CardMedia';
+import Button from '@material-ui/core/Button';
+import Typography from '@material-ui/core/Typography';
+import 'App.css';
+
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import * as MoviesActions from '../actions/actions';
```
Comment (zonzujiro) [inline src/components/MainPage.js]: The absolute path here.

Also, about importing of a `*` - usually it is bad for three shaking. So use it with wisdom :)

File: src/components/MainPage.js [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    margin: '20px auto',
+  },
+  media: {
+    height: 0,
+    paddingTop: '100%',
+  },
+});
+
+class MainPage extends Component {
+  constructor(props) {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: If you wish, you can declare `state`  without constructor like this:

`state = {/* */}`

File: src/components/MainPage.js [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+      error: null,
+      isLoaded: false,
+      inputValue: '',
+    };
+  }
+
+  componentDidMount() {
+    this.props.actions.fetchMovies();
+    this.setState({
+      isLoaded: true,
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Fetching of movies is async action, so you will set your flag `isLoaded` right after sending of the request.
Also, such flags must come from module thru `mapStateToProps`

File: src/components/MainPage.js [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    });
+  }
+
+  handleSearchChange = event => {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Pls, add types

File: src/components/MainPage.js [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+    this.setState({ inputValue: event.target.value });
+  };
+
+  handleSubmit = event => {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Types

File: src/components/MainPage.js [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+    event.preventDefault();
+    this.props.actions.searchMovies(this.state.inputValue);
+  };
+
+  render() {
+    const { error, isLoaded } = this.state;
+    const { classes } = this.props;
+    const data = this.props.data;
```
Comment (zonzujiro) [inline src/components/MainPage.js]: You can combine both of this lines

File: src/components/MainPage.js [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+    const poster = 'https://image.tmdb.org/t/p/w500';
+
+    if (error) {
+      return <div>Error: {error.message}</div>;
+    } else if (!isLoaded) {
+      return <div>Is loading...</div>;
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Pls, leave TODO that we need some `Loading` component and create an issue

File: src/components/MainPage.js [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+    } else {
+      return (
+        <Fragment>
```
Comment (zonzujiro) [inline src/components/MainPage.js]: So much `Fragment` 😄 You can use `div` already I think.

File: src/components/MainPage.js [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+          <Fragment>
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Redundant `Fragment`

File: src/components/MainPage.js [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+              </Card>
+            </div>
+          ))}
+        </Fragment>
+      );
+    }
+  }
+}
+
+MainPage.propTypes = {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Remove after adding of types

File: src/components/MainPage.js [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+  actions: PropTypes.object,
+  data: PropTypes.array,
+};
+
+function mapStateToProps(state) {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Why you need `state.defaultMovies`?

File: src/components/MainPage.js [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+  if (state.searchMovies.length > 0) {
+    return { data: state.searchMovies };
+  } else if (state.defaultMovies) {
+    return { data: state.defaultMovies };
+  }
+}
+
+function mapDispatchToProps(dispatch) {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: I think better will be if you use just use an object here:

```js
const mapDispathToProps = {
  /* list of actions */
}
```

In such way you will see what concrete actions passed to the container

File: src/reducers/defaultMoviesReducer.js [hunk 24]
```diff
@@ -0,0 +1,16 @@
+/* @flow */
+import initialState from './initialState';
```
Comment (zonzujiro) [inline src/reducers/defaultMoviesReducer.js]: Pls, move to module

File: src/reducers/defaultMoviesReducer.js [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+import { MOVIE_LOADING, MOVIE_LOADED } from '../actions/actionTypes';
```
Comment (zonzujiro) [inline src/reducers/defaultMoviesReducer.js]: Absolute path

File: src/reducers/defaultMoviesReducer.js [hunk 26]
```diff
@@ -0,0 +1,16 @@
...
+
+export const defaultMovies = (state = initialState.data, action) => {
```
Comment (zonzujiro) [inline src/reducers/defaultMoviesReducer.js]: Types?

**Review by zonzujiro — CHANGES_REQUESTED** [review]
Great work, keep going!

Pls, see the description of redux-ducks-modules. This architecture usually is used for applications with a redux.

File: src/components/MainPage.js [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+
+type State = {
+  inputValue: number | string,
+  error: Object,
+};
+
+class MainPage extends Component<Props, State> {
+  constructor(props) {
+    super(props);
+    this.state = {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: You have this data in store so you must just pass it from the redux store. No need to save the copy in the local state.

File: src/components/MainPage.js [hunk 28]
```diff
@@ -0,0 +1,142 @@
...
+    margin: '20px auto',
+  },
+  media: {
+    height: 0,
+    paddingTop: '100%',
+  },
+});
+
+type Props = {
+  defaultAction: Object,
```
Comment (zonzujiro) [inline src/components/MainPage.js]: `defaultAction` is a bad naming. Pls, rename to something obvious.

File: src/components/MainPage.js [hunk 29]
```diff
@@ -0,0 +1,142 @@
...
+      inputValue: '',
+      movies: [],
+    };
+  }
+
+  componentDidMount() {
+    this.props.defaultAction.getDefaultMovies();
+  }
+
+  handleSearchChange = (event: SyntheticEvent<any>) => {
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Pls, don't use `any`.  Or set concrete action or you can do something like this:

```js
handleSearchChange = ({ target }: { target: HTMLInputElement }) =>
```

In such way you will add type only for `target`

File: src/components/MainPage.js [hunk 30]
```diff
@@ -0,0 +1,142 @@
...
+    const poster = 'https://image.tmdb.org/t/p/w500';
+
+    if (error) {
+      return <div>Error: {error.message}</div>;
+    } else if (!data.isLoaded) {
+      return <LinearProgress color="secondary" variant="query" />;
+    } else {
+      return (
+        <div>
+          <Fragment>
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Pls, remove this `Fragment`, because it is redundant

File: src/components/MainPage.js [hunk 31]
```diff
@@ -0,0 +1,142 @@
...
+            </div>
+          ))}
+        </div>
+      );
+    }
+  }
+}
+
+function mapStateToProps(state) {
+  return { data: state.movies, isLoaded: true };
```
Comment (zonzujiro) [inline src/components/MainPage.js]: `isLoaded` also should be taken from the store

File: src/modules/searchMoviesReducer.js [hunk 32]
```diff
@@ -0,0 +1,40 @@
+/* @flow */
+import api from 'utils/Tmdb.js';
```
Comment (zonzujiro) [inline src/modules/searchMoviesReducer.js]: I think, that better will be if this module will be for all the movies, not only for some functionality as searching.  In such way, you will have fewer modules in the application.

So naming `movies` will fit better.

File: src/modules/searchMoviesReducer.js [hunk 33]
```diff
@@ -0,0 +1,40 @@
...
+const initialState = {
+  data: [],
+  isLoaded: false,
+  searchMovies: [],
+};
+
+const MOVIE_LOADING = 'MOVIE_LOADING';
+const MOVIE_LOADED = 'MOVIE_LOADED';
+
+export default function reducer(state: Object = initialState, action: Object) {
```
Comment (zonzujiro) [inline src/modules/searchMoviesReducer.js]: Pls, leave `TODO` that we need some concrete types for the state and for actions.

File: src/modules/searchMoviesReducer.js [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+export function receiveSearchMovies(json: Array<number | string> = []) {
+  return { type: MOVIE_LOADED, data: json, isLoaded: true };
+}
+
+export function searchMovies(value: string | number) {
+  return dispatch => {
+    const searchParams = {
+      query: value,
+    };
+    const movies = api.searchAmong('multi', searchParams);
```
Comment (zonzujiro) [inline src/modules/searchMoviesReducer.js]: `async/await`? 😄

File: src/components/MainPage.js [hunk 35]
```diff
@@ -0,0 +1,142 @@
...
+import CardContent from '@material-ui/core/CardContent';
+import CardMedia from '@material-ui/core/CardMedia';
+import Button from '@material-ui/core/Button';
+import Typography from '@material-ui/core/Typography';
+import LinearProgress from '@material-ui/core/LinearProgress';
+import 'App.css';
+
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import * as defaultMoviesReducer from 'modules/defaultMoviesReducer';
```
Comment (zonzujiro) [inline src/components/MainPage.js]: Why do you need `defaultMoviesReducer ` again?

**Review by zonzujiro — CHANGES_REQUESTED** [review]
Ok, final push. You can do it! 💪

File: .huskyrc [hunk 36]
```diff
@@ -1,5 +1,5 @@
 {
   "hooks": {
-    "pre-push": "echo 'husky: Running pre-push tasks...' && npm run prettify && npm run lint"
+    "pre-push": "echo 'husky: Running pre-push tasks...' && yarn prettify && yarn lint"
```
Comment (OleksiyRudenko) [inline .huskyrc:3]: Special thanks for making this consistent with project's package manager in use

**Review by zonzujiro — APPROVED** [review]
Great job!

**Review by OleksiyRudenko — APPROVED** [review]
Well done! Please, merge into master

**Comment by myroslav-kolomiiets** [comment]
Done. Thanks for review.

**Pull request merged**